lib/posix/pthread.c

```diff
@@ -656,6 +656,12 @@ int pthread_join(pthread_t pthread, void **status)
 		err = k_thread_join(&t->thread, K_FOREVER);
 		/* other possibilities? */
 		__ASSERT_NO_MSG(err == 0);
+
+		if (status != NULL) {
+			*status = t->retval;
+		}
+
+		posix_thread_recycle();
 	}
 
 	return 0;
```

Comment:
Looks like this should be `return ret` so that you get `ESRCH` or `EINVAL` when there's an error?

Reply:
Looks like https://github.com/zephyrproject-rtos/zephyr/pull/60964/files covers that. How about changing 641 and 644 to a pair of `k_spin_unlock(); return E...;` and then dropping the `if (ret == 0) {` and just handle the success case inline?

Reply:
Exactly - fixed in a separate PR

Reply:
> How about changing 641 and 644 to a pair of k_spin_unlock(); return E...;, inline the else case up there and then dropping the if (ret == 0) { and just handle the success case inline?

That sounds reasonable, let me give it another spin.

Reply:
@fabiobaltieri - how does this look?

Reply:
fancy...